### PR TITLE
Add streaming subprotocol negotiation

### LINKED ISSUE #15702: kubectl exec -i exits prematurely
With the current Git version of kubectl (see below), I cannot get an interactive shell in a container:

```
kubectl exec -ti single-container-pod bash
```
just exits after a couple of seconds, without echoing any TTY input.

Similarly,
```
cat large-file | kubectl exec -i single-container-pod -- bash -c "cat > /tmp/test && ls -l /tmp/test"
```
shows that in many cases, the file is not completely transferred.  "large-file" was 1MB in my case, with 6Mbit upload rate.

Client Version: version.Info{Major:"1", Minor:"2+",
GitVersion:"v1.2.0-alpha.1.1092+b8aabbbf4f92f0",
GitCommit:"b8aabbbf4f92f006cd1a118dbfd14d97bb996478",
GitTreeState:"clean"}

### LINKED ISSUE #15726: Add protocol versions & negotiation for exec, attach, port forward
As seen in #15702, trying to use kubectl from master (which includes #13322) against GKE (currently Kube 1.0.6, which does not include 13322) results in exec falling to function properly. We need to add protocol versions and negotiation to exec, attach, and port-forward to ensure backwards and forwards compatibility.

@smarterclayton @liggitt @lavalamp @kubernetes/kubectl @kubernetes/rh-ux @kubernetes/rh-cluster-infra

### PR DESCRIPTION
**Pull request by ncdc**
Add streaming subprotocol negotiation for exec, attach, and port
forwarding. Restore previous (buggy) exec functionality as an
unspecified/unversioned subprotocol so newer kubectl clients can work
against 1.0.x kubelets.

Fixes #15702
Fixes #15726

### CODE REVIEW

**Comment by ncdc** [comment]
This is my WIP for getting this implemented. I still need to flesh out some unit tests, and I'd like to restructure some of the exec/attach/portforward server code to make it easier to add new subprotocol version implementations, but that may have to wait depending on timing.

@smarterclayton @lavalamp @liggitt @kubernetes/rh-cluster-infra @bgrant0607

**Comment by lavalamp** [comment]
Made a quick skim; nothing stands out. Thanks for doing this!

**Comment by bronger** [comment]
The following is tested against current GKE.  The command

    kubectl exec -ti single_container_pod bash

quits without taking input with kubectl 1.1, however, with this PR, I can happily use the bash interactively.

Something like

    echo "Hello" | kubectl_old exec -i postgresql-vahuk 'bash -c "cat > /tmp/toll"'

hangs as with kubectl 1.0, but if I understand the PR correctly, this is hangxing as expected, isn't it?

**Comment by ncdc** [comment]
@bronger yes. This PR restores the previous 1.0 behavior when a newer kubectl connects to a 1.0.x kubelet. There is no way to eliminate that hang unless we backport the fix to 1.0.x. The hang is fixed with new kubectl + new kubelet.

**Comment by lavalamp** [comment]
@ncdc What needs to be done still before we can review & merge?

**Comment by ncdc** [comment]
@lavalamp I've been looking at refactoring exec & attach so the code is better able to support multiple protocol handlers internally, but I don't think that's necessary to fix this issue. I have 1 local commit that I haven't pushed yet that does the same type of refactoring for port forward, and I think that's ready to go assuming the units pass (I need to double-check). I'll push as soon as that's ready, and then we can move forward with this and defer the exec/attach changes to later, if that sounds good?

**Comment by ncdc** [comment]
@lavalamp updated, PTAL

**Comment by lavalamp** [comment]
@smarterclayton do you have time to review this or would you like me to take it?

**Comment by ncdc** [comment]
If it would make it simpler, I could back out the 3rd commit (creating a port forward protocol handler) as it's not strictly necessary to fix the original issue (and then bring it back in a follow-up).

**Comment by smarterclayton** [comment]
Let's split the third commit into another pull to streamline this.

**Comment by ncdc** [comment]
Ok, I'll save it off to the side, rebase to strip that out, and then start addressing your comments.

**Comment by ncdc** [comment]
I've got to step away for a couple of hours, but I will resume later this evening.

**Comment by ncdc** [comment]
Trying to figure out why doc verification is failing...

**Comment by ncdc** [comment]
Figured out docs issue. Fixing.

**Comment by ncdc** [comment]
To avoid an import cycle, I had placed a new constant in pkg/kubelet/server.go to be used by both the client and server port forwarding code. Unfortunately, this had the side effect of pulling in all kubelet-related flags into kubectl. I have moved the constant into a new subpackage under pkg/kubelet so the extra flags aren't pulled in. Eventually I will be moving protocol handlers and ServePortForward into this package as well.

**Comment by ncdc** [comment]
@smarterclayton ready for another round of review

**Comment by ncdc** [comment]
e2e flake... `[Fail] Pods [It] should not back-off restarting a container on LivenessProbe failure `

@k8s-bot test this

**Comment by smarterclayton** [comment]
What e2e test cover this?  Can we call the old protocol and verify it behaves a certain way from an e2e test somehow?

**Comment by ncdc** [comment]
We have e2e tests for exec:

- no stdin
- non-interactive stdin
- pseudo-interactive stdin

These all use the kubectl built with the release. There is no way currently to specify the subprotocol you want. The only way right now to test that is to run the kubectl from this branch against a 1.0.x kubelet...

**Comment by ncdc** [comment]
Additionally, for the pseudo-interactive stdin test case, as mentioned in #15726, it passes with a kubectl from master and a 1.0 kubelet, even though when you manually run the same test, it doesn't work. It appears to be a timing issue in the test itself, so I'm not sure of any reliable way to test that.

**Comment by smarterclayton** [comment]
Can you open an issue to have an integration test that mocks a backend that
verifies the client logic is correct?  We don't need it in 1.1 (kube or
openshift) but we will want it.

**Comment by ncdc** [comment]
Issue opened: #16119

File: pkg/util/httpstream/httpstream.go [hunk 1]
```diff
@@ -96,3 +102,44 @@ func IsUpgradeRequest(req *http.Request) bool {
...
+}
+
+// Handshake performs a subprotocol negotiation. If the client did not request
+// a specific subprotocol, defaultProtocol is used. If the client did request a
+// subprotocol, Handshake will select the first common value found in
+// serverProtocols. If a match is found, Handshake adds a response header
+// indicating the chosen subprotocol. If no match is found, HTTP forbidden is
+// returned, along with a response header containing the list of protocols the
+// server can accept.
+func Handshake(req *http.Request, w http.ResponseWriter, serverProtocols []string, defaultProtocol string) (string, error) {
```
Comment (smarterclayton) [inline pkg/util/httpstream/httpstream.go]: Unit test for this

File: pkg/client/unversioned/remotecommand/remotecommand.go [hunk 2]
```diff
@@ -97,155 +95,98 @@ func NewStreamExecutor(upgrader httpstream.UpgradeRoundTripper, fn func(http.Rou
...
-	errorChan := make(chan error)
-	headers.Set(api.StreamType, api.StreamTypeError)
-	errorStream, err := conn.CreateStream(headers)
+type streamProtocolHandler interface {
+	stream(httpstream.Connection) error
+}
+
+// Stream opens a protocol streamer to the server and streams until a client closes
+// the connection or the server disconnects.
+func (e *streamExecutor) Stream(stdin io.Reader, stdout, stderr io.Writer, tty bool) error {
```
Comment (smarterclayton) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: Unit test for this

**Comment by ncdc** [comment]
@smarterclayton unit test added for `httpstream.Handshake`. You also asked for a test for `Stream`... that's already partially covered by `TestRequestExecuteRemoteCommand`, although the only thing it tests w.r.t protocol negotiation is that the client & server are both using v2.channel.k8s.io. If you want a unit test that the client selects the correct `streamProtocolHandler`, I think `Stream` will need to be refactored a bit.

**Comment by smarterclayton** [comment]
This is LGTM for now.

**Comment by lavalamp** [comment]
@ncdc did/can you cherry-pick this?

**Comment by ncdc** [comment]
@lavalamp I have not cherry picked it, but we do need to pull it into 1.1. I won't have time to do it tonight, but can do it tomorrow morning if needed.

**Comment by lavalamp** [comment]
@ncdc Yes, please do, thanks!

**Comment by ncdc** [comment]
@lavalamp @smarterclayton in order to cherry pick, it looks like we'll need pull in #12283 (port forward backwards compat fixes) and #14618 (exec refactoring to use it without a client.Config) first. Is that ok?

**Comment by lavalamp** [comment]
@ncdc Yes, I think that's OK, thanks.

**Pull request merged**